src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -4414,6 +4414,212 @@ class StubGenerator: public StubCodeGenerator {
+  void wide_mul(Register prod_lo, Register prod_hi, Register n, Register m) {
+    __ mul(prod_lo, n, m);
+    __ mulhu(prod_hi, n, m);
+  }
+  void wide_madd(Register sum_lo, Register sum_hi, Register n,
+                 Register m, Register tmp1, Register tmp2) {
+    wide_mul(tmp1, tmp2, n, m);
+    __ cad(sum_lo, sum_lo, tmp1, tmp1);  // Add tmp1 to sum_lo with carry output to tmp1
+    __ adc(sum_hi, sum_hi, tmp2, tmp1);  // Add tmp2 with carry to sum_hi
+  }
+
+  // Poly1305, RFC 7539
+
+  // See https://loup-vaillant.fr/tutorials/poly1305-design for a
+  // description of the tricks used to simplify and accelerate this
+  // computation.
+
+  address generate_poly1305_processBlocks() {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "poly1305_processBlocks");
+    address start = __ pc();
+    Label here;
+
+    RegSet saved_regs = RegSet::range(x18, x25);
+    __ push_reg(saved_regs, sp);
+
+    RegSetIterator<Register> regs = RegSet::range(x10, x31).begin();
+
+    // Arguments
+    const Register input_start = *regs, length = *++regs, acc_start = *++regs, r_start = *++regs;
+
+    // R_n is the 128-bit randomly-generated key, packed into two
+    // registers. The caller passes this key to us as long[5], with
+    // BITS_PER_LIMB = 26.
+    const Register R_0 = *++regs, R_1 = *++regs;
+    pack_26(R_0, R_1, r_start);
+
+    const Register tmp1 = t0;
+    const Register tmp2 = t1;
+    const Register tmp3 = t2;
+
+    // RR_n is (R_n >> 2) * 5
+    const Register RR_0 = *++regs, RR_1 = *++regs;
+    __ srli(tmp1, R_0, 2);
+    __ shadd(RR_0, tmp1, tmp1, tmp2, 2);
+    __ srli(tmp1, R_1, 2);
+    __ shadd(RR_1, tmp1, tmp1, tmp2, 2);
+
+    // U_n is the current checksum
+    const Register U_0 = *++regs, U_1 = *++regs, U_2 = *++regs;
+    pack_26(U_0, U_1, U_2, acc_start);
+
+    static constexpr int BLOCK_LENGTH = 16;
+    Label DONE, LOOP;
+
+    __ addi(tmp1, zr, checked_cast<u1>(BLOCK_LENGTH));
+    __ blt(length, tmp1, DONE); {
+      __ bind(LOOP);
+
+      // S_n is to be the sum of U_n and the next block of data
+      const Register S_0 = *++regs, S_1 = *++regs, S_2 = *++regs;
+      __ ld(S_0, Address(input_start, 0));
+      __ ld(S_1, Address(input_start, wordSize));
+
+      __ cad(S_0, S_0, U_0, tmp1); // Add U_0 to S_0 with carry output to tmp1
+      __ cadc(S_1, S_1, U_1, tmp1); // Add U_1 with carry to S_1 with carry output to tmp1
+      __ adc(S_2, zr, U_2, tmp1);
+
+      __ addi(S_2, S_2, 1);
+
+      const Register U_0HI = *++regs, U_1HI = *++regs;
+
+      // NB: this logic depends on some of the special properties of
+      // Poly1305 keys. In particular, because we know that the top
+      // four bits of R_0 and R_1 are zero, we can add together
+      // partial products without any risk of needing to propagate a
+      // carry out.
+      wide_mul(U_0, U_0HI, S_0, R_0);  wide_madd(U_0, U_0HI, S_1, RR_1, tmp1, tmp2); wide_madd(U_0, U_0HI, S_2, RR_0, tmp1, tmp2);
```

Comment:
Fixed.

Reply:
It's written that way in the AArch64 code because it's (a more compact form of the mathematics) written that way in https://datatracker.ietf.org/doc/html/rfc7539#ref-Poly1305 . The point is to show the correspondence with the original mathematics, so that the reader can see what's happening.

In the paper, it's
```
r0·h0 + 5·2**−130·r1.h3 + 5·2**−130·r2·h2 + 5·2**−130·r3·h1,
r0·h1 +           r1·h0 + 5·2**−130·r2·h3 + 5·2**−130·r3·h2,
r0·h2 +           r1·h1 +           r2·h0 + 5·2**−130·r3·h3,
r0·h3 +           r1·h2 +           r2·h1 +           r3·h0
```
This is:
```
r0·h0 + 5·2**−130·r1.h3 + 5·2**−130·r2·h2,
r0·h1 +           r1·h0 + 5·2**−130·r2·h3,
r0·h2 +           r1·h1 +           r2·h0,
```